src/posts/xarray-kvikio/index.md

```diff
@@ -0,0 +1,145 @@
+
+### What is Kvikio
+
+> kvikIO is a Python library providing bindings to cuFile, which enables GPUDirectStorage (GDS).
+
+For Xarray, the key bit is that kvikio exposes a zarr store [kvikio.zarr.GDSStore](https://docs.rapids.ai/api/kvikio/stable/api.html#zarr) that does all the hard work for us. Since Xarray knows how to read Zarr stores, we can adapt that in a new storage backend. And thanks to recent work funded by the Chan Zuckerberg Initiative, adding a [new backend](https://docs.xarray.dev/en/stable/internals/how-to-add-new-backend.html) is quite easy!
+
+## Integrating with Xarray
+
+Getting all this to work nicely requires using three in-progress pull requests that
+
+1. [Teach Zarr to handle alternative array classes](https://github.com/zarr-developers/zarr-python/pull/934)
+2. [Rewrite a small bit of Xarray to not cast all data to a numpy array after read from disk](https://github.com/pydata/xarray/pull/6874)
+3. [Make a backend that connects Xarray to Kvikio](https://github.com/xarray-contrib/cupy-xarray/pull/10)
+
+Writing the backend for Xarray was relatively easily. Most of the code was copied over from the existing Zarr backend. Most of the effort was in ensuring that dimension coordinates could be read in directly to host memory without raising an error. This is required because Xarrays creates `pandas.Index` objects for such variables. In the future, we could consider using `cudf.Index` instead to allow a fully GPU-backed Xarray object.
+
+## Usage
+
+Assuming you have all the pieces together (see [Appendix I]() and [Appendix II]() for step-by-step instructions), then using all this cool technology only requires adding `engine="kvikio"` to your `open_dataset` line (!)
+
+```python
+import xarray as xr
+
+ds = xr.open_dataset("file.zarr", engine="kvikio", consolidated=False)
+```
+
+With this `ds.load()` will load directly to GPU memory and `ds` will now contain CuPy arrays.
+
+At present there are a few limitations:
+
+1. stores cannot be read with consolidated metadata, and
+2. compression is unsupported by the backend.
+
+## Quick demo
+
+First create an example uncompressed dataset to read from
+
+```
+import xarray as xr
+
+store = "./air-temperature.zarr"
+
+airt = xr.tutorial.open_dataset("air_temperature", engine="netcdf4")
+
+for var in airt.variables:
+    airt[var].encoding["compressor"] = None
+airt.to_zarr(store, mode="w", consolidated=True)
+```
+
+Now read
+
+```
+# consolidated must be False
+ds = xr.open_dataset(store, engine="kvikio", consolidated=False)
+ds
+```
+
+Now load a small subset
+
+```python
+type(ds["air"].isel(time=0, lat=10).load().data)
+```
+
+```
+cupy._core.core.ndarray
+```
+
+Success!
+
+Xarray integrates [decently well](https://cupy-xarray.readthedocs.io/quickstart.html) with CuPy arrays so you should be able to test out analysis pipelines pretty seamlessly.
+
+## Cool demo
+
+We don't have a cool demo yet but are looking to develop one very soon!
```

Comment:
Link to https://github.com/pangeo-data/xbatcher/issues/87?

Reply:
Sure, and/or we can link to the forum at https://discourse.pangeo.io/tag/machine-learning if people want to reach out.